src/game/client/baseheightmap.cpp

```diff
@@ -1233,122 +1233,141 @@ void BaseHeightMapRenderObjClass::Init_Dest_Alpha_LUT()
-    }
-
-    float z = pos1.z;
-    float z2 = (pos2.z - z) * (1.0f / i10);
-    unsigned char *data = map->Get_Data_Ptr();
-    int x_extent = map->Get_X_Extent();
-    int y_extent = map->Get_Y_Extent();
-
-    for (int y_pos = 0; y_pos < i10 && x_pos >= 0 && i2 >= 0 && x_pos < x_extent - 1 && i2 < y_extent - 1; y_pos++) {
-        int i11 = x_extent * i2 + x_pos;
-        float f1;
-
-        if (data[i11 + 1] >= data[i11]) {
-            f1 = data[i11 + 1];
+        x_increment_1 = 0;
+        y_increment_2 = 0;
+        major_distance = x_dist;
+        minor_distance = x_dist / 2;
+        minor_extent = y_dist;
+        major_extent = x_dist;
+    }
+
+    float current_height = pos1.z;
+    float height_increment = (pos2.z - current_height) * (1.0f / major_extent);
+    unsigned char *height_data = height_map->Get_Data_Ptr();
+    int x_extent = height_map->Get_X_Extent();
+    int y_extent = height_map->Get_Y_Extent();
+
+    // Iterate over the height map to check for obstructions
+    for (int i = 0; i < major_extent && current_x_pos >= 0 && current_y_pos >= 0 && current_x_pos < x_extent - 1
+         && current_y_pos < y_extent - 1;
+         i++) {
+        int map_index = x_extent * current_y_pos + current_x_pos;
+        float max_height_1;
+
+        // Find the maximum value among neighboring height map data (current, right, down, and diagonal)
+        if (height_data[map_index + 1] >= height_data[map_index]) {
+            max_height_1 = height_data[map_index + 1];
         } else {
-            f1 = data[i11];
+            max_height_1 = height_data[map_index];
         }
 
-        float f2;
+        float max_height_2;
 
-        if (data[x_extent + i11] >= f1) {
-            f2 = data[x_extent + i11];
+        if (height_data[x_extent + map_index] >= max_height_1) {
+            max_height_2 = height_data[x_extent + map_index];
         } else {
-            f2 = f1;
+            max_height_2 = max_height_1;
         }
 
-        float f3;
+        float max_height_3;
 
-        if (data[x_extent + 1 + i11] >= f2) {
-            f3 = data[x_extent + 1 + i11];
+        if (height_data[x_extent + 1 + map_index] >= max_height_2) {
+            max_height_3 = height_data[x_extent + 1 + map_index];
         } else {
-            f3 = f2;
+            max_height_3 = max_height_2;
         }
 
-        if (z + 0.5f < f3 * HEIGHTMAP_SCALE) {
+        // Check if there is an obstruction in the line of sight
+        if (current_height + 0.5f < max_height_3 * HEIGHTMAP_SCALE) {
             return false;
         }
 
-        if (Get_Max_Height() <= z && z2 > 0.0f) {
+        // Check if the viewer is looking upwards and the maximum height is reached
+        if (Get_Max_Height() <= current_height && height_increment > 0.0f) {
             return true;
         }
 
-        z = z + z2;
-        i8 += i9;
+        // Update the height and increment variables
+        current_height = current_height + height_increment;

```

Comment:
If you want you can also write

`current_height += height_increment;`

here.